torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,602 @@
+  auto& H_diag = state.H_diag();
+  auto& prev_flat_grad = state.prev_flat_grad();
+  auto& prev_loss = state.prev_loss();
+  auto& al = state.al();
+
+  int n_iter = 0;
   Tensor ONE = torch::tensor(1, flat_grad.options());
 
-  int64_t n_iter = 0;
-  while (n_iter < options.max_iter()) {
-    n_iter++;
-    state_n_iter++;
+  // optimize for a max of max_iter iterations
+  while(n_iter < max_iter) {
+    // keep track of nb of iterations
+    n_iter += 1;
+    state.n_iter(state.n_iter() + 1);
 
-    if (state_n_iter == 1) {
+    // compute gradient descent direction
+    if (state.n_iter() == 1) {
       d = flat_grad.neg();
       H_diag = ONE;
-      prev_flat_grad = flat_grad.clone(at::MemoryFormat::Contiguous);
     } else {
-      Tensor y = flat_grad.sub(prev_flat_grad);
-      Tensor s = d.mul(t);
-      Tensor ys = y.dot(s);
-
-      if (ys.item<float>() > 1e-10) {
+      // do lbfgs update (update memory)
+      auto y = flat_grad.sub(prev_flat_grad);
+      auto s = d.mul(t);
+      auto ys = y.dot(s); // y*s
+      if (val(ys) > 1e-10) {
         // updating memory
-
-        if (old_dirs.size() == options.history_size()) {
-          // shift history by one (limited memory)
+        if (int(old_dirs.size()) == history_size) {
+          // shift history by one (limited-memory)
           old_dirs.pop_front();
           old_stps.pop_front();
+          ro.pop_front();
         }
-
         // store new direction/step
-        old_dirs.push_back(y);
-        old_stps.push_back(s);
+        old_dirs.emplace_back(y);
+        old_stps.emplace_back(s);
+        ro.emplace_back(1. / ys);
 
         // update scale of initial Hessian approximation
-        H_diag = ys / y.dot(y);
+        H_diag = ys / y.dot(y);  // (y*y)
       }
 
-      int64_t num_old = old_dirs.size();
+      // compute the approximate (L-BFGS) inverse Hessian
+      // multiplied by the gradient
+      auto num_old = old_dirs.size();
 
-      for (int64_t i = 0; i < num_old; i++) {
-        ro.at(i) = ONE / old_dirs.at(i).dot(old_stps.at(i));
+      if (al.size() == 0) {
+        al.resize(history_size);
       }
 
-      Tensor q = flat_grad.neg();
-      for (int64_t i = num_old - 1; i >= 0; i--) {
+      // iteration in L-BFGS loop collapsed to use just one buffer
+      auto q = flat_grad.neg();
+      for (int i = num_old - 1; i >= 0; i--) {
         al.at(i) = old_stps.at(i).dot(q) * ro.at(i);
-        q.add_(old_dirs.at(i), -al.at(i).item());
+        q.add_(old_dirs.at(i), -val(al.at(i)));
       }
 
-      // Multiply by initial Hessian
+      // multiply by initial Hessian
       // r/d is the final direction
-      Tensor r = q.mul(H_diag);
-      d = r;
-
-      for (int64_t i = 0; i < num_old; i++) {
-        Tensor be_i = old_dirs.at(i).dot(r) * ro.at(i);
-        r.add_(old_stps.at(i), (al.at(i) - be_i).item());
+      d = torch::mul(q, H_diag);
+      for (size_t i = 0; i < num_old; i++) {
+        auto be_i = old_dirs.at(i).dot(d) * ro.at(i);
```

Comment:
```suggestion
        auto be_i = old_dirs.at(i).dot(r) * ro.at(i);
```
to be more consistent :D